File: swapPairs.cpp
//
// Created by Tengjun Gao on 11/29/21.
//

//给定一个链表，两两交换其中相邻的节点，并返回交换后的链表。
//
// 你不能只是单纯的改变节点内部的值，而是需要实际的进行节点交换。
//
//
//
// 示例 1：
//
//
//输入：head = [1,2,3,4]
//输出：[2,1,4,3]
//
//
// 示例 2：
//
//
//输入：head = []
//输出：[]
//
//
// 示例 3：
//
//
//输入：head = [1]
//输出：[1]
//
//
//
//
// 提示：
//
//
// 链表中节点的数目在范围 [0, 100] 内
// 0 <= Node.val <= 100
//
//
//
//
// 进阶：你能在不修改链表节点值的情况下解决这个问题吗?（也就是说，仅修改节点本身。）
// Related Topics 递归 链表 👍 1140 👎 0

#include <stdio.h>

using namespace std;
//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
  struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
  };
 */
struct ListNode {
    int val;
    ListNode *next;

    ListNode() : val(0), next(nullptr) {}

    ListNode(int x) : val(x), next(nullptr) {}

    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        // no swapping values
        ListNode *dummyHead = new ListNode(0); // 设置一个虚拟头结点
        dummyHead->next = head; // 将虚拟头结点指向head，这样方面后面做删除操作
        ListNode *cur = dummyHead;
        while (cur->next != nullptr && cur->next->next != nullptr) {
            ListNode *tmp = cur->next; // 记录临时节点
            ListNode *tmp1 = cur->next->next->next; // 记录临时节点

            cur->next = cur->next->next;    // 步骤一
            cur->next->next = tmp;          // 步骤二
            cur->next->next->next = tmp1;   // 步骤三

            cur = cur->next->next; // cur移动两位，准备下一轮交换
        }
        return dummyHead->next;
    }

    void printList(ListNode* head) {
        ListNode *ptr = head;
        while (ptr != nullptr) {
            printf("%d->", ptr->val);
            ptr = ptr->next;
        }
        printf("\n\n");
    }

};

//leetcode submit region end(Prohibit modification and deletion)
int main() {
    ListNode *node = new ListNode(1,
                                  new ListNode(2,
                                               new ListNode(3,
                                                            new ListNode(4))));
    Solution *s = new Solution();
    s->printList(node);
    ListNode *result = s->swapPairs(node);
    s->printList(result);
}

File: swapPairs.py
# Given a linked list, swap every two adjacent nodes and return its head. You
# must solve the problem without modifying the values in the list's nodes (i.e.,
# only nodes themselves may be changed.)
#
#
#  Example 1:
#
#
# Input: head = [1,2,3,4]
# Output: [2,1,4,3]
#
#
#  Example 2:
#
#
# Input: head = []
# Output: []
#
#
#  Example 3:
#
#
# Input: head = [1]
# Output: [1]
#
#
#
#  Constraints:
#
#
#  The number of nodes in the list is in the range [0, 100].
#  0 <= Node.val <= 100
#
#
#  Related Topics Linked List Recursion 👍 11414 👎 417
from typing import Optional


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head  # finish looping or linked list is empty

        # define previous, current, next node, change point direction
        pre = head
        cur = head.next
        next = head.next.next

        # switch current and pre
        cur.next = pre
        # assign next as another group of 3 nodes start with pre for swap
        pre.next = self.swapPairs(next)  # Note: assign pre.next to the previous result

        return cur


# leetcode submit region end(Prohibit modification and deletion)
if __name__ == '__main__':
    ex = ListNode(1,ListNode(2, ListNode(3, ListNode(4))))
    s = Solution()
    res = s.swapPairs(ex)
    pass